sodetlib/operations/bias_wave.py

```diff
@@ -0,0 +1,565 @@
+        high_current_mode (bool):
+            If true, switches to high-current-mode. If False, leaves in LCM
+            which runs through the bias-line filter, so make sure you
+            extend the step duration to be like >2 sec or something
+        hcm_wait_time (float):
+            Time to wait after switching to high-current-mode.
+        run_analysis (bool):
+            If True, will attempt to run the analysis to calculate DC params
+            and tau_eff. If this fails, the analysis object will
+            still be returned but will not contain all analysis results.
+        analysis_kwargs (dict, optional):
+            Keyword arguments to be passed to the BiasWaveAnalysis run_analysis
+            function.
+        channel_mask : np.ndarray, optional
+            Mask containing absolute smurf-channels to write to disk
+        g3_tag: string, optional
+            Tag to attach to g3 stream.
+        stream_subtype : optional, string
+            Stream subtype for this operation. This will default to 'bias_waves'.
+        enable_compression: bool, optional
+            If True, will tell the smurf-streamer to compress G3Frames. Defaults
+            to False because this dominates frame-processing time for high
+            data-rate streams.
+        plot_rfrac : bool
+            Create rfrac plot, publish it, and save it. Default is True.
+        show_plots : bool
+            Show plot in addition to saving when running interactively. Default is False.
+    """
+    if bgs is None:
+        bgs = cfg.dev.exp['active_bgs']
+    bgs = np.atleast_1d(bgs)
+
+    # Dumb way to get all run kwargs, but we probably want to save these in
+    # data object
+    freqs_wave = np.sort(np.atleast_1d(freqs_wave)) # Enforces lowest frequency first.
+    run_kwargs = {
+        'bgs': bgs, 'amp_wave': amp_wave,
+        'freqs_wave': freqs_wave, 'duration': duration,
+        'high_current_mode': high_current_mode,
+        'hcm_wait_time': hcm_wait_time, 'run_analysis': run_analysis,
+        'analysis_kwargs': analysis_kwargs, 'channel_mask': channel_mask,
+    }
+
+    initial_ds_factor = S.get_downsample_factor()
+    initial_filter_disable = S.get_filter_disable()
+    initial_dc_biases = S.get_tes_bias_bipolar_array()
+
+    try:
+        dc_biases = initial_dc_biases
+        init_current_mode = sdl.get_current_mode_array(S)
+        if high_current_mode:
+            dc_biases = dc_biases / S.high_low_current_ratio
+            amp_wave /= S.high_low_current_ratio
+            sdl.set_current_mode(S, bgs, 1)
+            S.log(f"Waiting {hcm_wait_time} sec after switching to hcm")
+            time.sleep(hcm_wait_time)
+
+        bwa = BiasWaveAnalysis(S, cfg, bgs, run_kwargs=run_kwargs)
+
+        bwa.sid = sdl.stream_g3_on(
+            S, tag=g3_tag, channel_mask=channel_mask, downsample_factor=1,
+            filter_disable=True, subtype=stream_subtype, enable_compression=enable_compression
+        )
+
```

Comment:
running each frequency on each BL for 3 seconds

Reply:
Depends on duration, and number of freqs arguments. Maybe @bkelle can give you the answer on the 2-frequencies and 3 sec duration test run we did and could probably do some scaling tests (for both time to run and data size) if we can get more time on satp3.

Reply:
It looks like yesterday for 2 frequencies it took about 2 mins! Which is basically consistent with bias steps then, ~1 min per frequency. I guess that means we should think about the data collection paradigm for i.e. time constant analysis with multiple frequencies as this could get large with like 7 frequencies or something like that.